pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+	if err != nil {
+		return
+	}
+	diff.add(from, to-1, value)
+}
+
+// Helper function to walk the array elements of type string.
+func walkArrayStrings(input string, visit func(from int, to int)) {
+	var scanner scanner
+	scanner.reset()
+	i, err := stepTo(&scanner, input, 0, scanBeginArray)
+	if err != nil {
+		return
+	}
+	stringFrom := -1
+	depth := 0
+	for ; i < len(input); i++ {
+		c := input[i]
+		switch scanner.step(&scanner, c) {
+		case scanBeginObject, scanBeginArray:
+			depth++
+		case scanEndObject:
+			depth--
+		case scanEndArray:
+			if depth == 0 {
+				return
+			}
+			depth--
+
+		case scanBeginLiteral:
+			if depth == 0 && input[i] == '"' {
+				stringFrom = i
+			}
+		case scanContinue:
+			if stringFrom != -1 && input[i] == '"' && input[i-1] != '\\' {
+				visit(stringFrom, i)
+				stringFrom = -1
+			}
+
+		case scanError:
+			return
+		}
+	}
+}
+
+// Helper function to walk the object keys and values.
+func walkObject(scanner *scanner, input string, i int, visit func(keyFrom, keyTo, valueFrom, valueTo int)) (int, error) {
+	i, err := stepTo(scanner, input, i, scanBeginObject)
+	if err != nil {
+		return i, err
+	}
+	keyFrom := -1
+	keyTo := -1
+	valueFrom := -1
+	depth := 0
+	for ; i < len(input); i++ {
+		switch scanner.step(scanner, input[i]) {
+		case scanBeginObject, scanBeginArray:
+			depth++
+		case scanEndArray:
+			depth--
+		case scanEndObject:
+			if depth != 0 {
+				depth--
+				continue
+			}
+			// We reached the end of the object we were scanning
+			if keyFrom != -1 && keyTo != -1 && valueFrom != -1 {
+				// Visit the last value of the objet
+				visit(keyFrom, keyTo, valueFrom, i)
+			}
+			return i + 1, nil
+
+		case scanBeginLiteral:
+			if depth != 0 || keyFrom != -1 {
+				continue
+			}
+			if input[i] == '"' {
+				keyFrom = i
+			}
+		case scanContinue:
+			if keyFrom != -1 && keyTo == -1 && input[i] == '"' && input[i-1] != '\\' {
+				keyTo = i + 1
+			}
+
+		case scanObjectKey:
+			if depth == 0 {
+				valueFrom = i + 1
+			}
+		case scanObjectValue:
+			if depth != 0 {
+				continue
+			}
+			visit(keyFrom, keyTo, valueFrom, i)
+			keyFrom = -1
+			keyTo = -1
+			valueFrom = -1
+
+		case scanError:
+			return i, scanner.err
+		}
+	}
+	scanner.eof()
+	return i, scanner.err
+}
+
+// Helper function to step to the given expected scanner operation `to`.
+func stepTo(scanner *scanner, input string, i int, to int) (int, error) {
+	for ; i < len(input); i++ {
+		switch op := scanner.step(scanner, input[i]); op {
+		default:
+			return i + 1, unexpectedScannerOpError(op)
+		case scanSkipSpace, scanContinue:
+			continue
+		case scanError:
+			return i + 1, scanner.err
+		case to:
+			return i + 1, nil
+		}
+	}
+	scanner.eof()
+	return i, scanner.err
+}
+
+// Helper function to step to one of the given expected scanner operations `to`.
+// An error is returned if another operation is found.
+func stepToOneOf(scanner *scanner, input string, i int, to ...int) (j int, op int, err error) {
```

Comment:
I think you can just use this one everywhere and get rid of `skipTo`. You can just call this one `skipTo` or `skipToOneOf` or `skipToAny`

Reply:
I distinguished them for performance reasons: the OneOf version is more expensive to call due to its variadic arguments and extra return value.

Reply:
Interesting. Have you actually benchmarked this? Do you have the numbers? I'd be curious to know.